### PR TITLE
Update GTK stack for use in CI/CD

### PR DESCRIPTION
**Pull request by Lord-Kamina**
Ok @cas--, you win; I'll split it.

I also opted to make a new PR  that will supersede #454, to make it easier to keep track of the discussion, were there to be any.

After this, I will make another PR for unpinning pytest (I _might_ include the change from `gtkreactor` to `gireactor` in this latter one because I genuinely think it would make everybody's lives easier, but you tell me)

This is just the bit updating the GTK stack. As you can see, I've reverted the changes to libtorrent and such (although I maintain it should be somewhat of a priority to address this sooner rather than later)

### CODE REVIEW

**Comment by Lord-Kamina** [comment]
@cas-- anything that needs to change here before this can be merged? (The CI fails should be resolved by triggering a manual run at the `gvsbuild-release` repo)

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -36,6 +42,7 @@ jobs:
 
       - name: Install dependencies
         run: |
+          sudo apt-get install libcairo2-dev libgirepository1.0-dev
```
Comment (cas--) [inline .github/workflows/ci.yml:45]: This doesn't seem right to be installing these dev dependencies, everything should have been built already?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -74,7 +81,8 @@ jobs:
     runs-on: windows-2022
     strategy:
       matrix:
-        python-version: ["3.7", "3.10"]
+        python-version: ["3.7", "3.9", "3.10"]
```
Comment (cas--) [inline .github/workflows/ci.yml:84]: Why are we including 3.9 here and removing it from package build?

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -77,7 +88,6 @@ jobs:
         working-directory: deluge_src
         run: |
           python -m pip install .
-          python setup.py install_scripts
```
Comment (cas--) [inline .github/workflows/cd.yml:80]: Are you sure removing this is correct?

File: .github/workflows/cd.yml [hunk 4]
```diff
@@ -24,8 +27,8 @@ jobs:
     strategy:
       matrix:
         arch: [x64, x86]
-        python: ["3.9"]
-        libtorrent: [2.0.7, 1.2.19]
+        python: ["3.7", "3.10"]
```
Comment (cas--) [inline .github/workflows/cd.yml:30]: Why do we need 3.7 build?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -10,13 +10,19 @@ on:
...
     strategy:
       matrix:
-        python-version: ["3.7", "3.10"]
-
+        python-version: ["3.9", "3.10"]
+        os: ["ubuntu-24.04"]
+        include:
+          - os: ubuntu-22.04
+            python-version: 3.7
+    runs-on: ${{ matrix.os }}
```
Comment (cas--) [inline .github/workflows/ci.yml:25]: Let's keep these Linux CI changes separate from the Windows GTK changes

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -89,6 +97,24 @@ jobs:
           cache: "pip"
           cache-dependency-path: "requirements*.txt"
 
+      - name: Determine gvsbuild release URL
```
[the remark this answers is not in the document]
Comment (Lord-Kamina) [reply .github/workflows/ci.yml:100]: The idea is to enable the gtk tests though. I think that might have gotten cut onto another PR though.

File: .github/workflows/docs.yml [hunk 7]
```diff
@@ -29,7 +29,7 @@ jobs:
         run: |
           pip install --upgrade pip wheel
           pip install tox
-          sudo apt-get install enchant-2
+          sudo apt-get install enchant-2 libgirepository1.0-dev
```
[the remark this answers is not in the document]
Comment (Lord-Kamina) [reply .github/workflows/docs.yml:32]: I honestly can't remember 100% but it is possible that it failed somewhere without it. Would have to look it up.

File: requirements-tests.txt [hunk 8]
```diff
@@ -1,4 +1,4 @@
-libtorrent
+libtorrent==2.0.7
```
[the remark this answers is not in the document]
Comment (Lord-Kamina) [reply requirements-tests.txt:1]: Yes because things in newer libtorrent made tests fail, so until tests are updated, it has to be.

File: .github/workflows/cd.yml [hunk 9]
```diff
@@ -24,8 +27,8 @@ jobs:
...
+        libtorrent: [2.0.8, 1.2.19]
```
[the remark this answers is not in the document]
Comment (Lord-Kamina) [reply .github/workflows/cd.yml:31]: Actually I think this was a typo or possibly I missed this when reverting. I can't find it now so it should go back to 2.0.7 (for now)

File: .github/workflows/cd.yml (hunk 4 above)
Comment (Lord-Kamina) [reply .github/workflows/cd.yml:30]: We don't really but I had kept it because it's what deluge was using and/or officially targetting at the moment.

The point here was to have a 3.7 build to represent not breaking the current build, alongside building with newer versions.

Comment (cas--) [reply .github/workflows/cd.yml:29]: The actual release only uses one version. There is a difference between CI tests that need to both test oldest and latest support dependencies versus this workflow that creates the Windows release package which should aim to be most stable recent release of dependencies.

Comment (Lord-Kamina) [reply .github/workflows/cd.yml:29]: I just noticed that the packaging test was not running at all on my fork (probably because I never had an event there that would trigger it) so I'll need a bit to check that everything is actually ok on that front.

I'm in the process of looking into some of your other comments and corrections. Also, please take a look at what I said you over on discord.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (Lord-Kamina) [reply .github/workflows/ci.yml:25]: That'll just prolong this whole thing. The linux changes are just: keep 22.04 (which needs an older python) and add a build on a newer version of ubuntu.

File: .github/workflows/cd.yml (hunk 4 above)
Comment (Lord-Kamina) [reply .github/workflows/cd.yml]: I'm leaving just x64-py3.10 but I put LT1.2 back in due to the outburst in Discord.

File: .github/workflows/cd.yml (hunk 3 above)
Comment (Lord-Kamina) [reply .github/workflows/cd.yml]: Nope. I think it might be with something I did on another branch but I'll cross that bridge when I get there.
Reverted for now.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Lord-Kamina) [reply .github/workflows/ci.yml:83]: I'm including it due to some differences in the importlib module. It's not relevant now but I think it will be to test against with some PR I've yet to submit.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Lord-Kamina) [reply .github/workflows/ci.yml]: I removed it for now.

File: .github/workflows/cd.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (Lord-Kamina) [reply .github/workflows/cd.yml]: Changed back to 2.0.7

File: .github/workflows/docs.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lord-Kamina) [reply .github/workflows/docs.yml]: Removed as well.

**Comment by Lord-Kamina** [comment]
BTW @cas--  I have pinned twisted here because of the issue with the maybe_coroutine test, until we can figure out what changed/is going on.
Other than that, I think most comments have been addressed.

File: requirements.txt [hunk 10]
```diff
@@ -1,5 +1,5 @@
 libtorrent
-twisted[tls]>=17.1
+twisted[tls]>=17.1,<=24.7
```
Comment (cas--) [inline requirements.txt:2]: I have resolved the problem with Twisted 24.10 so this can be removed

File: .github/workflows/cd.yml [hunk 11]
```diff
@@ -49,12 +52,20 @@ jobs:
...
       - name: Prepare pip
-        run: python -m pip install wheel setuptools==68.*
+        run: python -m pip install wheel setuptools==70.*
+
+      - name: Determine gvsbuild release URL
+        id: gvsbuild-url
+        shell: bash
+        run: |
+          test -z "${{ github.event.inputs.gvsbuild-tag }}" && tag=latest || tag="${{ github.event.inputs.gvsbuild-tag }}"
+          if [[ "$tag" == "latest" ]]; then URL="https://github.com/${{ github.repository_owner }}/gvsbuild-release/releases/$tag/download"; else URL="https://github.com/${{ github.repository_owner }}/gvsbuild-release/releases/download/$tag" ; fi
```
Comment (cas--) [inline .github/workflows/cd.yml:62]: This should be able to split over multiple lines to make it more readable?

File: .github/workflows/ci.yml (hunk 5 above)
Comment (cas--) [reply .github/workflows/ci.yml:25]: Can we keep the scope of this PR to Win GTK only please

File: .github/workflows/ci.yml (hunk 2 above)
Comment (cas--) [reply .github/workflows/ci.yml:83]: Let's leave that to the other PR then

File: .github/workflows/ci.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (cas--) [reply .github/workflows/ci.yml:99]: If we are not using it let's leave it out of this PR, the code is always going to be in CD.yml. Also it doesn't address the Linux GTK tests

File: requirements-tests.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (cas--) [reply requirements-tests.txt:1]: So for CI we already have this pinned so this only affects local testing, perhaps there shouldn't be a difference but that is also a question about pinning the entire test/dev stack to known working dependencies. I really do want to fix the tests for libtorrent > 2.0.7 so keeping it unpinned serves as a reminder

**Review by cas-- — CHANGES_REQUESTED** [review]
Can we please keep PRs to a single feature/fix as outlined here: https://testing.googleblog.com/2024/07/in-praise-of-small-pull-requests.html?m=1

**Comment by Lord-Kamina** [comment]
> Can we please keep PRs to a single feature/fix as outlined here: https://testing.googleblog.com/2024/07/in-praise-of-small-pull-requests.html?m=1

So here's my problem with this...

I understand the concept and motivation for wanting small PRs, and IMO I've been nothing if not accommodating. But other issues with it aside, this "nothing" PR has been open for months. This is not a dig on you or anyone else, because I know everyone working on this project does so as a volunteer freely contributing their time.

My point is this: These commits were meant as the first-step in getting a more modern CI pipeline as the foundation of other rather large and potentially breaking changes to deluge that require using newer versions of python (this is, dropping `pkg_resources` entirely and moving the plug-in system to using wheels and importlib).
That code is also already written and tested (rebasing, cleaning-up and addressing other issues aside).

And, that itself is just a preparatory step for something I've not yet done, which is make a better and easily-reproducible process for creating macOS packages and eventually, adding macOS to the CD.

With this in mind, and considering the snail pace at which this project moves (again, not a dig, but rather a fact), at some point you maybe gotta start wondering why don't you make it easier for everyone, yourself included? This was originally a much larger PR that involved fixing tests and also updating the python runtimes, which is why the there is an update on the linux workflow as well. I already split it and excluded the fixes to the testing suite, but at this point it's gone well beyond "small PRs are good". The change to the linux workflow is absolutely immaterial to the working of deluge or the CI, and is directly related to the intent of this PR.

It's incredibly frustrating because it seems like we're gettint fixated with dividing what should have been a simple task into an ever-increasing number of steps ad nauseam.

So, to recap: I understand and fully support the spirit of advocating for small PRs, but I think you're being maybe a tad excessive on this matter.
